# Import Libraries

In [2]:
# Importing libraries (pandas, numpy, os)
import pandas as pd
import numpy as np
import os

# Import Data

In [3]:
# Create a path
path = r'C:\Users\Owner\Documents\Instacart Basket Analysis'

In [4]:
# Import products.csv file using a path
df_products = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import orders.csv file using a path
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Consistency Checks for df_products

In [6]:
#Create a dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix']=['a','b',1,True]

In [8]:
df_test

mix
0     a
1     b
2     1
3  True

In [9]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
    
# The structure that’s being used in this code is called a “for-loop.”
# The “for” in for-loop stands for “for these elements, do this,”
# and the “loop” describes how the structure works: looping over and over again as it performs the procedures detailed by the “for.”
# Here, the for-loop is looping through each column in the dataframe and executing the same block of code each time.    
    
# Within the for-loop, a new variable is created: weird.
# Assigned to it is a test that checks whether the data types within the column are consistent. 
# The weird variable will ultimately take a boolean value of either True or False. 
# If True, that means the column contains inconsistent data types. 
# If False, that means the column contains only one data type.
# Boolean values can also be represented by numbers: 0 as False and 1 as True.

# Here comes the “if” statement.
# An if statement checks if some condition is met, and if it’s met, executes a line of code.
# If the condition isn’t met, the code isn’t executed.
# Here, the if statement is checking whether weird is true or false.
# If it’s greater than 0, than it’s true. If not, it’s false.
# If weird is true, the command print(col) is executed, which prints the problematic column for you to see.
# Because of the for-loop, this command will be executed on every column in your dataframe, printing every mixed-type column it finds.

mix


# Check for missing values

In [10]:
#looking for missing values
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
#Let's look at the 16 missing values
df_nan = df_products[df_products['product_name'].isnull()==True]

In [12]:
# view the missing values subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Mean and Median

In [13]:
#descriptive data
df_products.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [14]:
#what is the median of the data set
df_products.median()

C:\Users\Owner\AppData\Local\Temp\ipykernel_30228\2442910398.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_products.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [15]:
df_products.mean()

C:\Users\Owner\AppData\Local\Temp\ipykernel_30228\1786336332.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_products.mean()


product_id       24844.345139
aisle_id            67.770249
department_id       11.728433
prices               9.994136
dtype: float64

In [16]:
# Replacing missing value with mean
## df['column with missings'].fillna(mean value, inplace=True)

In [17]:
# Replacing missing value with median
## df['column with missings'].fillna(median value, inplace=True)

In [18]:
#How many rows and columns of data?
df_products.shape

(49693, 5)

In [19]:
# Create dataframe without missing values in 'product_name' column
df_products_clean = df_products[df_products['product_name'].isnull() == False]

In [20]:
df_products_clean.shape

(49677, 5)

49,693 - 49,677 = 16
These are the 16 rows of missing data

In [21]:
# Check again for mixed-type data after filtering out missing values
for col in df_products_clean.columns.tolist():
  weird = (df_products_clean[[col]].applymap(type) != df_products_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_products_clean[weird]) > 0:
    print (col)

# Duplicates

In [22]:
df_dups = df_products_clean[df_products_clean.duplicated()]

In [23]:
# locating duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# Before doing so, let’s check the current number of rows in your df_prods_clean dataframe so that you can compare the number after removing the duplicates

df_products_clean.shape

(49677, 5)

In [25]:
# Creating a new dataframe that doesn’t include the duplicates
# df_prods_clean_no_dups that contains only the unique rows from df_prods_clean.

df_products_clean_no_dups = df_products_clean.drop_duplicates()

In [26]:
#run the new code to see if duplicates have been deleted.
df_products_clean_no_dups.shape

(49672, 5)

49,677 - 49,672 = 5. The five duplicates have been successfully deleted.

# Exercise 4.5 

1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [27]:
df_orders.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

There is nothing odd about the Orders descriptive data. The first column is unnamed. The second column is order ID. The third column is user ID. The fourth column is order number. The order Days of Week is from 0 to 6, that's normal. The hours of the day are from 0 to 23, that's normal. The days since prior order is from 0 to 30, that's normal.

3. Check for mixed-type data in your df_ords dataframe.

In [28]:
for col in df_orders.columns.tolist():
  weird = (df_orders[[col]].applymap(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
order_dow  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

I did not find any mixed-type data in the df_orders dataframe. All the results returned as "consistent".

5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [29]:
df_orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [30]:
# Drop unnamed column for df_ords

df_orders = df_orders.drop(columns = ['Unnamed: 0'])

In [31]:
df_orders.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [32]:
# Drop eval_set column for df_ords

df_orders = df_orders.drop(columns = ['eval_set'])

In [33]:
df_orders.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [34]:
df_orders.shape

(3421083, 6)

In [35]:
# Create subset for the missing values in df_ords

df_orders_nan = df_orders[df_orders['days_since_prior_order'].isnull() == True]

In [36]:
# View the missing values subset
df_orders_nan

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
11        2168274        2             1          2                 11   
26        1374495        3             1          1                 14   
39        3343014        4             1          6                 11   
45        2717275        5             1          3                 12   
...           ...      ...           ...        ...                ...   
3420930    969311   206205             1          4                 12   
3420934   3189322   206206             1          3                 18   
3421002   2166133   206207             1          6                 19   
3421019   2227043   206208             1          1                 15   
3421069   3154581   206209             1          3                 11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

The only missing values appear in the "Days since prior order" column. This makes sense because when an order is first generated, it wouldn't have a previous order since it's a new order.

6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

The only missing values were in the "Days since prior order" column. It would not be appropriate to change those dates. So therefore, I didn't change the missing values.

7. Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [37]:
df_orders_dups = df_orders[df_orders.duplicated()]

In [38]:
df_orders_dups.shape

(0, 6)

No duplicates found. 

8. Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

I didn't find any duplicates, therefore, I didn't change any of the data.

9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [39]:
# Export changes: 
df_orders.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked_v2.csv'))

In [40]:
# Export changes: 
df_products.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [41]:
df_orders.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [42]:
df_orders.shape

(3421083, 6)